Add comprehensive documentation for Microsoft Entra Identity Sidecar (#3535)
AzureAD/microsoft-identity-web: docs/sidecar/agent-identities.md

# Agent Identities

> Agent identities enable sophisticated authentication scenarios where an agent application acts autonomously or on behalf of specific users. This document explains the semantics and usage patterns for agent identities in the sidecar.

Reviewer: ```suggestion
Agent identities enable sophisticated authentication scenarios where an agent application acts autonomously or on behalf of users for interactive agents. This document explains the semantics and usage patterns for agent identities in the sidecar. It focuses on the SDK aspects. More documentation exists for agents identity in general in [Agentic identity platform documentation](https://learn.microsoft.com/entra/agentic-identity-platform)
```

## Overview

Agent identities support two primary patterns:

1. **Autonomous Agent**: The agent application operates in its own application context

> 2. **Delegated Agent**: The agent application operates on behalf of a specific user identity

Reviewer: ```suggestion
2. **Delegated Agent**: An interactive agent that operates on behalf the user that triggered it.
```

- Batch operations on behalf of multiple users
- Systems using object IDs for user reference

### Pattern 4: Regular Request (No Agent)

When no agent parameters are provided, the sidecar uses the incoming token's identity.

**Scenario**: Standard on-behalf-of (OBO) flow without agent identities.

```bash
GET /AuthorizationHeader/Graph
```

**Token Characteristics**:
- Token type: Depends on incoming token and configuration
- Uses standard OBO or client credentials flow
- No agent identity facets

## Precedence Rules

When both `AgentUsername` and `AgentUserId` are somehow present (though validation should prevent this), the implementation gives precedence to username:

```
AgentUsername > AgentUserId
```

However, this scenario should be avoided and will result in a validation error in strict validation mode.

## Microsoft Entra ID Configuration

### Prerequisites for Agent Identities

1. **Agent Application Registration**:
   - Register the parent agent application in Microsoft Entra ID
   - Configure API permissions for downstream APIs
   - Set up client credentials (certificate or secret)

2. **Agent Identity Configuration**:
   - Create agent identity registrations

>    - Configure Federated Identity Credentials (FIC) between the parent agent and agent identities

Reviewer: ```suggestion
```
this is done by the registration No need to do that

   - Assign necessary permissions to agent identities

3. **Application Permissions**:
   - Grant application permissions for autonomous scenarios
   - Grant delegated permissions for user delegation scenarios
   - Ensure admin consent is provided where required

### Example: Configuring an Agent Identity

```bash
# Create agent identity app registration
az ad app create --display-name "MyAgent Identity"

# Get the agent identity client ID
AGENT_IDENTITY_ID=$(az ad app list --display-name "MyAgent Identity" --query [0].appId -o tsv)

# Configure FIC between parent agent and agent identity
az ad app federated-credential create \
  --id $PARENT_AGENT_APP_ID \
  --parameters '{
    "name": "MyAgentIdentityFIC",
    "issuer": "https://login.microsoftonline.com/<tenant-id>/v2.0",
    "subject": "'$AGENT_IDENTITY_ID'",
    "audiences": ["api://AzureADTokenExchange"]
  }'

# Grant permissions to the agent identity
az ad app permission add \
  --id $AGENT_IDENTITY_ID \
  --api 00000003-0000-0000-c000-000000000000 \
  --api-permissions e1fe6dd8-ba31-4d61-89e7-88639da4683d=Scope

# Grant admin consent
az ad app permission admin-consent --id $AGENT_IDENTITY_ID
```

## Token Claims

> ### Autonomous Agent Token Claims

Reviewer: We need to remove this example and point to the (future) Agentic identity platform documentation

string? parentAgent = claimsPrincipal.GetParentAgentBlueprint();
```

## Error Scenarios

### Missing AgentIdentity with AgentUsername

**Request**:
```bash
GET /AuthorizationHeader/Graph?AgentUsername=user@contoso.com
```

**Response**:
```json
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "Bad Request",
  "status": 400,
  "detail": "AgentUsername requires AgentIdentity to be specified"
}
```

### Both AgentUsername and AgentUserId Specified

**Request**:
```bash
GET /AuthorizationHeader/Graph?AgentIdentity=agent-id&AgentUsername=user@contoso.com&AgentUserId=user-oid
```

**Response**:
```json
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "Bad Request",
  "status": 400,
  "detail": "AgentUsername and AgentUserId are mutually exclusive"
}
```

### Invalid AgentUserId Format

**Request**:
```bash
GET /AuthorizationHeader/Graph?AgentIdentity=agent-id&AgentUserId=invalid-guid
```

**Response**:
```json
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "Bad Request",
  "status": 400,
  "detail": "AgentUserId must be a valid GUID"
}
```

## Code Examples

### TypeScript/JavaScript

> ```typescript

Reviewer 2: Agent perspective: Also need to show what happens when it FAILS

Current (from comments): Examples show happy path only

Agent-friendly example:

```TypeScript
// TypeScript - Agent calling sidecar
async function getTokenForUser(userToken: string) {
  try {
    const response = await fetch('http://sidecar:8080/AuthorizationHeader', {
      method: 'POST',
      headers: { 
        'Content-Type': 'application/json',
        'Authorization': userToken  // Forward user token
      },
      body: JSON.stringify({
        AgentIdentity: process.env.AGENT_ID,
        Scopes: ['User.Read']
      })
    });
    
    if (!response.ok) {
      const error = await response.json();
      
      // Agent can self-correct based on error code
      if (error.errorCode === 'AGENT_001') {
        console.error('Missing required AgentIdentity parameter');
        // Retry with correct structure
      }
      
      throw new Error(`Sidecar error: ${error.message}`);
    }
    
    return await response.json();
    
  } catch (err) {
    // Handle network errors, timeouts, etc.
    console.error('Failed to reach sidecar:', err);
    throw err;
  }
}
```
Why: An agent reading this learns:

- How to make the request
- How to handle errors
- How to retry with corrections
- What to log for debugging
